### PR TITLE
40/second

### PR DESCRIPTION
**Pull request by ChrisForti**
- Everything tests out, including the desired state tests.
- Now to write a test for the repository configuration

### CODE REVIEW

**Comment by ChrisForti** [comment]
- What does it mean when i run a code block in my terminal and it gives me a redirection prompt?
```
chris@docker:~/self-hosting-on-raspi $ if (which docker)
then
  echo "Docker already installed"
else
  echo "Installing docker"
  sudo apt install docker-ce docker-ce-cli containerd.io docker-buildx-plugin d>fi
> 
> 
> 
> 
> 
> 
> 
> 
>  
```

**Comment by ChrisForti** [comment]
- Output from `sudo docker run hello-world`
```
chris@docker:~/self-hosting-on-raspi $  sudo docker run hello-world
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
478afc919002: Pull complete 
Digest: sha256:a26bff933ddc26d5cdf7faa98b4ae1e3ec20c4985e6f87ac0973052224d24302
Status: Downloaded newer image for hello-world:latest

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (arm64v8)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/
```

**Comment by ChrisForti** [comment]
-  `tee` - read from standard input and write to standard output and files....
- Not sure if the `|` and `tee` need to be on that line?

**Comment by ChrisForti** [comment]
- Tested on a fresh image, and the ca-certificates did not fire off, because I was testing with `which`
- CA certificates are the small digital files that make digital trust, digital identity, and security possible—whether across the internet or within your company network.
- for now i used `test -d`, but I think the `stat` command would be better, just have to find the right flag, arg combo.

File: dockers-install.sh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  sudo chmod a+r /etc/apt/keyrings/docker.gpg
+fi
+
+# Intializing repository, and placing copies in .list files
+if (stat -c "%n" /etc/os-release && echo "$VERSION_CODENAME")
+then 
+  echo "Repository already initialized"
+else
+  echo "Initializing repo"
+  echo \
```
Comment (stephenlauck) [inline dockers-install.sh]: Maybe just get used to not breaking lines, I think it's harder to read and confusing when there are a bunch of `\`, but others disagree.

File: dockers-install.sh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
```
Comment (stephenlauck) [inline dockers-install.sh:62]: what is `. /etc/os-release` doing?

File: dockers-install.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+if (stat -c "%a" /etc/bin/keyrings/docker.asc)
+then
+  echo "gpg keyring already exists"
+else
+  echo "Installing gpg keyring"
+  sudo curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
+fi
+ 
+# setting permissions 
+if (test -f /etc/bin/keyrings/docker.gpg)
```
Comment (stephenlauck) [inline dockers-install.sh]: This isn't the test for permissions, maybe try and figure out a test for a+r?

File: dockers-install.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+if (test -d /etc/bin/keyrings)
+then
+  echo "creating keyrings"
+else
+  echo "Creating keyrings"
+  sudo install -m 0755 -d /etc/apt/keyrings
+fi
+
+# Adding a gpg keyring
+if (stat -c "%a" /etc/bin/keyrings/docker.asc)
```
Comment (stephenlauck) [inline dockers-install.sh]: This check would be maybe ok with just `gpg /etc/bin/keyrings/docker.asc` and gpg will check it's validity? I'm not sure but try it.

**Review by stephenlauck — CHANGES_REQUESTED** [review]

File: dockers-install.sh (hunk 2 above)
Comment (ChrisForti) [reply dockers-install.sh:62]: - It is a susbshell that starts with a path to the file that contains the Operating system type && echoing an env. variable which I believe is identifying the operating system release code name.

File: dockers-install.sh (hunk 4 above)
Comment (ChrisForti) [reply dockers-install.sh]:
```
chris@docker:~ $ gpg /etc/bin/keyrings/docker.asc
gpg: directory '/home/chris/.gnupg' created
gpg: keybox '/home/chris/.gnupg/pubring.kbx' created
***gpg: WARNING: no command supplied.  Trying to guess what you mean ...***
gpg: can't open '/etc/bin/keyrings/docker.asc'
```
- I tried that the other day, but the warning scared me away from it.

Comment (stephenlauck) [reply dockers-install.sh]: It's saying there is no file at that path. That is a good check. Then you should do an else that creates that file like you had.

Comment (ChrisForti) [reply dockers-install.sh]:
```
chris@docker:~ $ if ( gpg /etc/bin/keyrings/docker.asc)
then
  echo "gpg keyring already exists"
else
  echo "Installing gpg keyring"
  sudo curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
fi
gpg: WARNING: no command supplied.  Trying to guess what you mean ...
gpg: can't open '/etc/bin/keyrings/docker.asc'
Installing gpg keyring
```

File: dockers-install.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+then 
+  echo "Repository already initialized"
+else
+  echo "Initializing repo"
+  echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+fi 
+
+# Update
+sudo apt update
```
Comment (stephenlauck) [inline dockers-install.sh]: This is a hard one, but you should be checking when/if you need to run apt update. You can check if docker is in the apt-cache and it's the right version by using something like:

`apt-cache show docker` and if not in the apt cache, then do a `sudo apt update`.

**Review by stephenlauck — APPROVED** [review]

File: dockers-install.sh (hunk 5 above)
Comment (ChrisForti) [reply dockers-install.sh]: Thank you sir, I was not sure how to go about that.

Comment (ChrisForti) [reply dockers-install.sh]: I will run some test when I get in front of my machine.

**Pull request merged**